proposals/lockout-violation-detection.md

```diff
@@ -1,56 +1,96 @@
 ### Lockout Violation Detection
 
-We assume a database that we will store vote information ingested from
+Assume:
+1. A database that we will store vote information ingested from
 gossip/turbine, which in turn will be consulted to detect lockout violations
 as follows.
+2. First we track all SlotHashes on the canonical/rooted fork in the database
+for the last epoch.
 
-1) Track all SlotHashes on the main fork in the database.
-2) If there are any votes made that have a root `R` missing from the SlotHashes
-list, slash them on this fork
-3) For all other slots `S` in each vote other than the root, track a range
+The following sections will go over the possible lockout violations that will
+be detected and punished.
+
+#### Rooting a different fork
+If there are any votes made that have a root `R` missing from the rooted
+SlotHashes list, slash them on the rooted fork since they have committed
+to a different fork
+
+#### Removing lockouts
+Attempting to illegally *remove* a lockout in a newer vote that that should
+have still existed based on an older vote is a lockout violation. We detect
+this as follows:
+
+1. For all other slots `S` in each vote other than the root, track a range
 `(S, S + 2^n)` where `n` is the confirmation count. For each slot `S` we only
 have to track the greatest such lockout range in the database.
-4) For each new vote `V`, for each slot `S` in the vote, lookup in the database
+
+2. For each new vote `V`, for each slot `S` in the vote, lookup in the database
 to see if there's some range `(X, X + 2^n)` where `S` is in this range, but
-`X < S` and `X` is missing from the vote `V`. This is a lockout violation
-because this implies that the validator made a vote where `S` popped off `X`,
-but the lockout on `X` from an earlier vote should have prevented that from
-happening.
+`X` is missing from the vote `V`. This is a lockout violation because this
+implies that the validator made a vote where `S` popped off `X`, but the
+lockout on `X` from an earlier vote should have prevented that from happening.
 
 Note for each interval `(S, S + 2^n)` we also need to be able to lookup the
```

Comment:
Might be out of scope for this proposal, but should we decide how to dedup slashing detections? For a new vote `V` there might be multiple offending intervals / past votes, should we be slashing each offense?

Reply:
yeah i think we slash each offending vote as a tuple (v1, v2), (v1, v3), etc. If (v2, v3) also conflict then triple slashing